lib/auth/predicate/predicate.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package predicate
+
+import (
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/trace"
+	"github.com/vulcand/predicate"
+)
+
+// NamedParameter is an object with a name that can be added to the environment in which a predicate expression runs.
+type NamedParameter interface {
+	// GetName returns the name of the object.
+	GetName() string
+}
+
+func buildEnv(objects ...NamedParameter) map[string]any {
+	env := make(map[string]any)
+	for _, obj := range objects {
+		env[obj.GetName()] = obj
+	}
+
+	return env
+}
+
+func newParser(env map[string]any) (predicate.Parser, error) {
+	getIdentifierInEnv := func(selector []string) (any, error) {
+		return getIdentifier(env, selector)
+	}
+
+	return predicate.NewParser(predicate.Def{
+		Operators: predicate.Operators{
+			AND: builtinOpAnd,
+			OR:  builtinOpOr,
+			NOT: builtinOpNot,
+			EQ:  builtinOpEquals,
+			LT:  builtinOpLT,
+			GT:  builtinOpGT,
+			LE:  builtinOpLE,
+			GE:  builtinOpGE,
+		},
+		Functions: map[string]any{
+			"add":            builtinAdd,
+			"sub":            builtinSub,
+			"mul":            builtinMul,
+			"div":            builtinDiv,
+			"xor":            builtinXor,
+			"split":          builtinSplit,
+			"upper":          builtinUpper,
+			"lower":          builtinLower,
+			"contains":       builtinContains,
+			"first":          builtinFirst,
+			"append":         builtinAppend,
+			"array":          builtinArray,
+			"replace":        builtinReplace,
+			"len":            builtinLen,
+			"regex":          builtinRegex,
+			"matches":        builtinMatches,
+			"contains_regex": builtinContainsRegex,
+			"map_insert":     builtinMapInsert,
+			"map_remove":     builtinMapRemove,
```

Comment:
I recently added support for "methods" within gravitational/predicate [here](https://github.com/gravitational/predicate/pull/4) which could provide support for insert and remove on different types basically through interfaces, without needing prefixes on the names. I think it's worth consideration if you weren't aware of them